state/protocol/protocol_state.go

```diff
@@ -92,142 +92,65 @@ type ProtocolState interface {
 	//
 	// Error returns:
-	//   - Per convention, the input seals from the block payload have already confirmed to be protocol compliant.
+	// [TLDR] All error returns indicate potential state corruption and should therefore be treated as fatal.
+	//   - Per convention, the input seals from the block payload have already been confirmed to be protocol compliant.
 	//     Hence, the service events in the sealed execution results represent the honest execution path.
 	//     Therefore, the sealed service events should encode a valid evolution of the protocol state -- provided
 	//     the system smart contracts are correct.
 	//   - As we can rule out byzantine attacks as the source of failures, the only remaining sources of problems
-	//     can be (a) bugs in the system smart contracts or (b) bugs in the node implementation.
-	//     A service event not representing a valid state transition despite all consistency checks passing
-	//     is interpreted as case (a) and handled internally within the StateMutator. In short, we go into Epoch
-	//     Fallback Mode by copying the parent state (a valid state snapshot) and setting the
-	//     `InvalidEpochTransitionAttempted` flag. All subsequent Epoch-lifecycle events are ignored.
+	//     can be (a) bugs in the system smart contracts or (b) bugs in the node implementation. A service event
+	//     not representing a valid state transition despite all consistency checks passing is interpreted as
+	//     case (a) and _should be_ handled internally by the respective state machine. Otherwise, any bug or
+	//     unforeseen edge cases in the system smart contracts would result in consensus halt, due to errors while
+	//     evolving the protocol state.
 	//   - A consistency or sanity check failing within the StateMutator is likely the symptom of an internal bug
 	//     in the node software or state corruption, i.e. case (b). This is the only scenario where the error return
 	//     of this function is not nil. If such an exception is returned, continuing is not an option.
-	ApplyServiceEventsFromValidatedSeals(seals []*flow.Seal) error
-}
-
-// DeferredBlockPersistOp is a database update function which is dependent on a candidate block ID.
-// Internally to the protocol_state package we don't have access to the candidate block ID yet because
-// we are still determining the protocol state ID for that block.
-type DeferredBlockPersistOp func(tx *transaction.Tx, blockID flow.Identifier) error
-
-// DeferredBlockPersistOps is a wrapper around `transaction.DeferredDBUpdate` which additionally
-// supports deferring database operations which depend on the not-yet-determined candidate block ID.
-// Once the protocol state for the candidate block is built, we can compute the candidate block ID and use
-// `Decorate` to contextualize the deferred operations with the appropriate block ID.
-type DeferredBlockPersistOps struct {
-	innerUpdates []DeferredBlockPersistOp
-}
```

Comment:
moved this logic in package `storage/badger/transaction` ([here](https://github.com/onflow/flow-go/blob/5a8a45f4d8eb712a708883f0d45f1410f59b31ed/storage/badger/transaction/deferred_block_persist.go)) and refactored it to be similar to [`DeferredDBUpdate`](https://github.com/onflow/flow-go/blob/5a8a45f4d8eb712a708883f0d45f1410f59b31ed/storage/badger/transaction/deferred_update.go)